src/Android/GliderLink.cpp

```diff
@@ -0,0 +1,131 @@
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+#include "GliderLink.hpp"
+#include "Util/StaticString.hxx"
+#include "org_xcsoar_GliderLinkReceiver.h"
+#include "Compiler.h"
+#include "Components.hpp"
+#include "Blackboard/DeviceBlackboard.hpp"
+#include "Context.hpp"
+
+Java::TrivialClass GliderLink::gl_cls;
+jmethodID GliderLink::gl_ctor_id, GliderLink::close_method;
+
+
+bool
+GliderLink::Initialise(JNIEnv *env)
+{
+  assert(!gl_cls.IsDefined());
+  assert(env != nullptr);
+
+  gl_cls.Find(env, "org/xcsoar/GliderLinkReceiver");
+
+  gl_ctor_id = env->GetMethodID(gl_cls, "<init>",
+                                 "(Landroid/content/Context;I)V");
+  close_method = env->GetMethodID(gl_cls, "close", "()V");
+
+  return true;
+}
+
+void
+GliderLink::Deinitialise(JNIEnv *env)
+{
+  gl_cls.Clear(env);
+}
+
+GliderLink* GliderLink::create(JNIEnv* env, Context* context,
+                                         unsigned index) {
+  assert(gl_cls != nullptr);
+
+  // Construct InternalGPS object.
+  jobject obj =
+    env->NewObject(gl_cls, gl_ctor_id, context->Get(), index);
+  assert(obj != nullptr);
+
+  GliderLink *glider_link = new GliderLink(env, obj);
+  env->DeleteLocalRef(obj);
+
+  return glider_link;
+}
+
+GliderLink::GliderLink(JNIEnv* env, jobject obj)
+    : obj(env, obj) {
+}
+
+GliderLink::~GliderLink() {
+  JNIEnv *env = Java::GetEnv();
+  env->CallVoidMethod(obj.Get(), close_method);
+}
+
+gcc_visibility_default
+JNIEXPORT void JNICALL
+Java_org_xcsoar_GliderLinkReceiver_setGliderLinkInfo(
+    JNIEnv* env, jclass cls, jint index, jlong gid, jstring callsign,
+    jdouble latitude, jdouble longitude, jdouble altitude,
+    jdouble gspeed, jdouble vspeed, jint bearing) {
+
+  ScopeLock protect(device_blackboard->mutex);
+  NMEAInfo &basic = device_blackboard->SetRealState(index);
+  basic.UpdateClock();
+  basic.alive.Update(basic.clock);
+
+  GliderLinkTrafficList &traffic_list = basic.glink_data.traffic;
+
+  GliderLinkId id = GliderLinkId((uint32_t)gid);
+
+  GliderLinkTraffic *traffic = traffic_list.FindTraffic(id);
+  if (traffic == nullptr) {
+    traffic = traffic_list.AllocateTraffic();
+    if (traffic == nullptr)
+      // no more slots available
+      return;
+
+    traffic->Clear();
+    traffic->id = id;
+
+    traffic_list.new_traffic.Update(basic.clock);
+  }
+
+  const char *nativeString = env->GetStringUTFChars(callsign, JNI_FALSE);
+  traffic->name.SetASCII(nativeString);
+  env->ReleaseStringUTFChars(callsign, nativeString);
+
+  traffic->location_available = true;
+  traffic->location = GeoPoint(Angle::Degrees(longitude),
+                              Angle::Degrees(latitude));
+
+  traffic->altitude_available = altitude > -10000.0;
+  traffic->altitude = altitude;
+  traffic->speed_received = gspeed >= 0.1;
+  traffic->speed = gspeed;
+  traffic->climb_rate_received = vspeed > -8675309.0;
+  traffic->climb_rate = vspeed;
+  traffic->track_received = bearing < 361.0;
```

Comment:
So this allows a track of `360.999999` degrees. Is this what you wanted? Why does it allow a track of `-99999999` degrees?